llama-index-core/llama_index/core/agent/workflow/base_agent.py

```diff
@@ -421,38 +420,88 @@ async def parse_agent_output(
             )
 
         if not ev.tool_calls:
-            memory = await ctx.store.get("memory")
             messages = await memory.aget()
             output = await self.finalize(ctx, ev, memory)
```

Comment:
the bug is right here. At least for the function calling agent, it first writes to a scratchpad, and then moves the scratchpad values to the memory only when finalize is called.

So, the fix here should just be to fetch the `messages =` assignment right after this line, rather than right before